### PR TITLE
Fix issues with createElement and attributeChangedCallback

### PR DESCRIPTION
**Pull request by bundyo**
Avoid repeatedly replacing createElement (causes issues with nested custom elements).
Avoid calling attributeChangedCallback when old and new values are the same (not really changed).

### CODE REVIEW

**Comment by WebReflection** [comment]
I'm afraid you probably got it wrong. The way the build system works won't allow any re-assignment.

If you are using directly the src/document-register-element.js file you might as well notice this on top:

```js
// DO NOT USE THIS FILE DIRECTLY, IT WON'T WORK
// THIS IS A PROJECT BASED ON A BUILD SYSTEM
// THIS FILE IS JUST WRAPPED UP RESULTING IN
// build/document-register-element.js
// and its .max.js counter part
```

while if you use the [build/document.register-element.max.js] file or its minified counter part (without max) you'll notice this. on top

```js
if (REGISTER_ELEMENT in document) return;
```

There is no case where the register element is addressed multiple times indeed.

If you need any other clarification please don't hesitate to ask

**Comment by WebReflection** [comment]
+ also you cannot drop the usage of `getAttribute` it's there for a reason.

Last, but not least, if there are no tests when there are changes, no PR will be accepted.

Please bear in mind this project is already used in production, give it a spin before pushing changes ;-)

Help/hints appreciated anyway, and for instance the "do not notify if same value as before" is IIRC not how standard work so don't set attributes repeatedly for no reason if you don't want that callback to trigger.

If you can link me the line in the standard documentation that states it shouldn't though, I'll be happy to fix the issue on a new bug.

**Comment by bundyo** [comment]
I meant that registerElement is called multiple times - once for each custom element you need. Each time it replaces document.createElement and for me this causes build/document-register-element.js to stop working completely - createdCallback never gets called. This happens both in Firefox and IE11 (since Chrome has native custom elements support). Of course I may be wrong and the reason to be completely different. I'll try to isolate a demo.

I wasn't sure about the attribute changes, as the MutationObserver fires events for same value changes too, but I've just tested the native Chrome custom elements implementation and looks like it doesn't fire attributeChangedCallback for same values. In any way, the draft doesn't say anything about that, and **changed** definition doesn't mention custom elements as it is outside of the draft, so it is probably your call to decide.

**Comment by WebReflection** [comment]
I don't know what you mean. The polyfill does lazy assignement but only once.

This is because of the amount of logic and changes involved to the DOM so until you register an element there is no overhead.

Feel free to test this here:
http://webreflection.github.io/document-register-element/test/

Also feel free to use this script in your page and write here what you think is not working:
http://cdnjs.cloudflare.com/ajax/libs/document-register-element/0.4.1/document-register-element.js

Once again, if you change or use anything inside the `src/` folder I tell you gonna have problems because that is **not** what you use in production, that is part of this repo as partial JS template.

If you have an example though, I'll be happy to verify the problem.

**Comment by WebReflection** [comment]
Last, to give you extra info of why I believe you got it wrong

// set as enumerable, writable and configurable
document[REGISTER_ELEMENT] = function registerElement(type, options) {
  upperType = type.toUpperCase();
  if (!setListener) {
    // only first time document.registerElement is used
    // we need to set this listener
    // setting it by default might slow down for no reason
    setListener = true;
```

that is the very same block createElement is reassigned ... **once** like I've said.

Please note this script is used in production from big sites too so something like that would have instantly caught ( but again I don't usually do these kind of things after all these years that would be quite lame from me ;-) )

Best Regards

**Comment by bundyo** [comment]
Okay, you've got me - it doesn't happen because of multiple reassignments of document.createElement - there is something else, which gets fixed when the createElement assignment is before the registerElement call...

Here is a demo:
http://bundyo.org/custom-elements/button.html

**Comment by WebReflection** [comment]
FF doesn't work here. I am not sure I understand what you are doing, you wrapped the poly and it gets on my way (in terms of debugging).

Can you please use plain `CustomElements` logic ?

You should also `registerElement` **before** `creatingElement` or just register it and let the body be revolved by itself.

Feel free to use this page as example: http://webreflection.github.io/document-register-element/test/examples/x-map.html

**Comment by WebReflection** [comment]
actually, I think I know what is your problem. You should read how to extend native here:
https://github.com/WebReflection/document-register-element#htmltablerowinputselectotherselement

Also please read every other related documentation in the README, buttons are part of the same INPUT/SELECT family.

TL;DR There is no way to extend them in a meaningful manner, you need the `is` attribute ;-)

**Comment by WebReflection** [comment]
```js
document.registerElement(
  'tp-button',
  {
    extends: 'button', // <== IMPORTANT
    prototype: Object.create(
      HTMLButtonElement.prototype,
      {
        createdCallback: {value: function () {
          this.classList.add(this.nodeName.toLowerCase());
        }}
      }
    )
  }
);
```

on the body
```html
<button is="tp-button"></button>
```

**Comment by bundyo** [comment]
Okay, my bad, let me fix that for you. :) I've changed it to inherit a simple HTMLElement - this shouldn't be a problem, right? Still doesn't work.

I meant above that FF works with its native custom elements implementation - when you switch on dom.webcomponents.enabled in about:config. Otherwise - no.

**Comment by bundyo** [comment]
Uploaded the non-minified polyfill, refresh for better debugging.

**Comment by WebReflection** [comment]
uhm ... actually you know what? something went terribly wrong at some point between 0.3.0 and 0.3.1 ... pleaase try again using this version instead:
cdnjs.cloudflare.com/ajax/libs/document-register-element/0.3.0/document-register-element.js

and let me know if that works as expected. I have no idea why nobody noticed this already, I need to understand how badly fucked up is this.

**Comment by bundyo** [comment]
Yep, it works :)

**Comment by WebReflection** [comment]
OK, so should `0.4.3` which I just pushed ... my deepest apologies for the inconvenient ... I just gave a talk about testing every bloody thing and the latest innerHTML didn't test nodes already in place.

It was a little, but potentially catastrophic bug, I'm actually glad you filed a fix for something else and tested for real. That's the spirit!!!!

Thank You! Please give the CDN few days to propagate, then feel free to use cdnjs.cloudflare.com/ajax/libs/document-register-element/0.4.3/document-register-element.js

**Comment by WebReflection** [comment]
And just because I've mentioned it, here the description of the `innerHTML` issue with the helper link included:
https://github.com/WebReflection/document-register-element#using-innerhtml

**Comment by WebReflection** [comment]
what a timing ... https://github.com/WebReflection/document-register-element/issues/28

**Comment by bundyo** [comment]
Thanks! :)

Btw, I mentioned to TJ about the issue today and he probably decided to test because of me :)
